Add targets for writing GDS files

lpawelcz opened the pull request:
Closes https://github.com/google/xls/issues/1024

This PR bumps `bazel_rules_hdl` in order to:
* Add `gds_write` rule for writing GDS files (see https://github.com/hdl/bazel_rules_hdl/pull/169).
* Add support for ASAP7 PDK (see https://github.com/hdl/bazel_rules_hdl/pull/176)

Additionally:
* https://github.com/google/xls/pull/1031/commits/65a89c31bf9d1d8adedb325dff9e24c807a5fe36 adds target which builds GDS files for `find_index` example based on the results from previous `place_and_route` targets for ASAP7 and SKY130 technologies.
* https://github.com/google/xls/pull/1031/commits/b67d8b02404d1895789ecf7d1bd378c6135fae95 adds targets for generating GDS files for recently added `RLE` encoder and decoder examples, also for ASAP7 and SKY130.

EDIT:
This PR depends on https://github.com/google/xls/pull/1230

lpawelcz commented:
PR is ready for review, I've updated the description. @proppy please take a look.

proppy commented on dependency_support/embedded_python_interpreter/print_libpython_path.py:
```diff
@@ -23,4 +23,13 @@
     for pv in ('LIBDIR', 'LIBPL')
 ]
 
+if 'rules_hdl_cpython' in config_vars['prefix']:
```
is there a reference bug for this workaround?

lpawelcz replied:
It looks like we can remove this workaround.

proppy commented on WORKSPACE:
```diff
@@ -73,3 +73,7 @@ initialize_external_repositories()
 load("@com_github_grpc_grpc//bazel:grpc_extra_deps.bzl", "grpc_extra_deps")
 
 grpc_extra_deps()
+
```
I suspect this is related to:
> It was necessary to add and register cpython repository because without it, bazel_rules_hdl would fail to install its pip dependencies

This is surprising that this is required here, since XLS also install pip deps from https://github.com/google/xls/blob/main/dependency_support/pip_requirements.txt

lpawelcz replied:
It looks like we can remove this workaround.

proppy commented on xls/examples/BUILD:
```diff
@@ -562,6 +563,13 @@ place_and_route(
+    density_fill_config = "fill.json"
+)
+
+gds_write(
```
should we add more rules for other examples?

proppy replied:
can we replace `_asic` with `_gds` or `_layout`?

proppy replied:
can we rename the rule?
`find_index_gds`
or 
`find_index_layout`

lpawelcz replied:
I missed that somehow, It's now fixed

proppy commented on xls/examples/fill.json:
```diff
@@ -0,0 +1,83 @@
+{
```
is that design specific or can it be reused across many design? also why is it not referenced from the bazel rules?

proppy replied:
if it's PDK specific I think we can maintain it in the PDK integration in bazel_rules_hdl, similar to how we do it for tcl files for synthesis and sta.

lpawelcz replied:
Yes, I think we can do that. I will work on adding this in bazel_rules_hdl

lpawelcz replied:
For the record, here is the link to relevant PR in bazel_rules_hdl: https://github.com/hdl/bazel_rules_hdl/pull/173

proppy commented on xls/examples/klayout.lyt:
```diff
@@ -0,0 +1,158 @@
+<?xml version="1.0" encoding="utf-8"?>
```
I think the actual upstream is https://github.com/efabless/sky130_klayout_pdk/blob/main/sky130_tech/tech/sky130/sky130.lyt

lpawelcz replied:
This is not a part of this PR anymore. It was added here only for testing purpose on the early stage of development.
It could be removed because KLayout tech files can now be supplied in `bazel_rules_hdl` with https://github.com/hdl/bazel_rules_hdl/commit/5e2ec27916ed3a5b956f76d63e632c24b8bd1f81

proppy commented on xls/examples/BUILD:
```diff
@@ -562,6 +563,13 @@ place_and_route(
+gds_write(
+	name = "find_index_asic",
+	implemented_rtl = ":find_index_place_and_route",
+	klayout_lyt = "klayout.lyt",
```
can we target asap7 instead?

lpawelcz replied:
ASAP7 support in `bazel_rules_hdl` is not entirely working. It looks like it was tested only to the synthesis stage. When I tried to use synthesis results generated with this PDK in rules `place_and_route` and `gds_write` I encountered some problems, see https://github.com/hdl/bazel_rules_hdl/issues/172

lpawelcz replied:
Please let me know if it would be OK for you to first merge this PR with the support for SKY130. Then once I solve the problems with ASAP7 I could open a follow up PR with ASAP7 examples.

proppy replied:
oh I realize that's probably outdated.

lpawelcz commented:
https://github.com/hdl/bazel_rules_hdl/pull/173 is merged. I bumped bazel_rules_hdl dependency, rebased the code and fixed the commit history. @proppy  this PR is ready for review again, please take a look.

proppy commented on xls/modules/rle/BUILD:
```diff
@@ -97,6 +101,35 @@ xls_ir_verilog(
+)
+
+gds_write(
+	name = "rle_enc_asic",
```
can we replace `_asic` with `_gds` or `_layout`?

lpawelcz replied:
changed to `rle_enc_gds`

lpawelcz commented:
@proppy  Sorry for the delay, I addressed your comments: renamed the rules and updated PR description.

proppy reviewed (approved):
just a nit.

lpawelcz commented:
There are some fuzz test timeouts in CI which don't not look related to our changes

lpawelcz commented:
Rebased the code, bumped bazel_rules_hdl further to include changes from https://github.com/hdl/bazel_rules_hdl/pull/176 effectively adding support for ASAP7 PDK. Renamed example targets to showcase which PDK is used in those and added examples for ASAP7. I also updated the PR description to reflect the current state of the PR. @proppy please take a look

proppy commented:
> There are some fuzz test timeouts in CI which don't look related to our changes

The fuzz timeout should be gone with https://github.com/google/xls/commit/47a51dd95f1e7a8c91104b4c6ccd209e8087fe2b can you rebase again?

proppy commented on xls/examples/BUILD:
nit: Load statement is out of its lexicographical order.

lpawelcz replied:
fixed

proppy commented on xls/modules/rle/BUILD:
nit: Load statement is out of its lexicographical order.

lpawelcz replied:
fixed

lpawelcz commented:
Fixed the load statements order and rebased the code onto https://github.com/google/xls/commit/4c05344b2e1e2a0815e1be66b4a1a017de4f618d

proppy commented:
Note: the internal review is currently blocked by a bazel_rules_hdl import (#ref 566360201)

lpawelcz commented:
Builds fail on GDS generation with:
```
  File "/home/runner/.cache/bazel/_bazel_runner/5bd26e4beeff37176a5bf6fac1c974c7/sandbox/linux-sandbox/27899/execroot/com_google_xls/bazel-out/k8-opt-exec-2B5CBBC6/bin/external/rules_hdl/gds_write/def2stream.runfiles/rules_hdl_pip_deps_klayout/site-packages/klayout/tl/__init__.py", line 1, in <module>
    import klayout.tlcore
ModuleNotFoundError: No module named 'klayout.tlcore'
```

This is caused by https://github.com/google/xls/commit/eb2fb6d81523b09fbd77b63b88381b1ad6c74850. I was able to fix the build locally by changing the python version from `3.9` to `3.8` but it's not a good solution. I believe this could be fixed by https://github.com/hdl/bazel_rules_hdl/issues/209

proppy commented:
https://github.com/google/xls/commit/eb2fb6d81523b09fbd77b63b88381b1ad6c74850 does pass the python interpreter target here: https://github.com/google/xls/blob/main/dependency_support/initialize_external.bzl#L37 but also set the version to `3.9` here:  https://github.com/google/xls/blob/main/dependency_support/rules_hdl/workspace.bzl#L53

Do you have more logging from bazel when `pip_parse` get executed?
https://github.com/hdl/bazel_rules_hdl/blob/c7da44c0e96a2b76b95e9b5d58b0272b8aa14796/init.bzl#L60-L67

lpawelcz commented:
Hi @proppy, I was actually able to fix this issue by fixing hermetic python handling in `bazel_rules_hdl`. I've opened a PR there: https://github.com/hdl/bazel_rules_hdl/pull/240. Please take a look at it.

In this PR for the sake of testing my changes I've added a temporary commit which changes the source of bazel_rules_hdl to branch from the linked PR. I will change it to bumping bazel_rules_hdl version once https://github.com/hdl/bazel_rules_hdl/pull/240 is merged.
